source/opt/isolate_function_calls_pass.h

```diff
@@ -0,0 +1,55 @@
+// Copyright (c) 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef SOURCE_OPT_ISOLATE_FUNCTION_CALLS_PASS_H_
+#define SOURCE_OPT_ISOLATE_FUNCTION_CALLS_PASS_H_
+
+#include <algorithm>
+#include <map>
+#include <queue>
+#include <unordered_map>
+#include <unordered_set>
+#include <utility>
+
+#include "source/opt/basic_block.h"
+#include "source/opt/def_use_manager.h"
+#include "source/opt/ir_context.h"
+#include "source/opt/module.h"
+#include "source/opt/pass.h"
+
+namespace spvtools {
+namespace opt {
+
+// See optimizer.hpp for documentation.
+class IsolateFunctionCallsPass : public Pass {
+ public:
+  IsolateFunctionCallsPass();
+  const char* name() const override { return "isolate-function-calls"; }
+  Status Process() override;
+
+  IRContext::Analysis GetPreservedAnalyses() override {
+    return IRContext::kAnalysisInstrToBlockMapping |
```

Comment:
I don't think the instr-to-block mapping is kept up-to-date.  I might be a good idea to keep the def-use manager up-to-date as well.  That is expensive to rebuild.  The gains you make in shaders with small basic block might be lost if the def-use chain need to be rebuilt.